core/src/main/java/org/apache/iceberg/rest/responses/FetchPlanningResultResponse.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest.responses;
+
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.rest.PlanStatus;
+import org.apache.iceberg.rest.RESTResponse;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface FetchPlanningResultResponse extends RESTResponse {
+  PlanStatus planStatus();
+
+  @Nullable
+  List<String> planTasks();
+
+  @Nullable
+  List<FileScanTask> fileScanTasks();
+
+  @Nullable
+  List<DeleteFile> deleteFiles();
+
+  @Override
+  default void validate() {
+    Preconditions.checkArgument(
+        planStatus() != null, "invalid, status can not be null: ", planStatus());
+    Preconditions.checkArgument(
+        planStatus() != PlanStatus.COMPLETED && (planTasks() != null || fileScanTasks() != null),
+        "invalid response, tasks can only be returned in a 'completed' status");
```

Comment:
Why does this throw `IllegalArgumentException`? Should this be some other REST exception because the service sent an invalid response? It isn't an argument that was the problem, it was the response itself.

Reply:
Not sure I follow but I think it can be considered an `IllegalArgumentException` since planStatus and tasks can be considered as arguments to the request builder, and if the combination of arguments are logically in a state where it's not spec compliant I feel like it makes sense to fail here? Another compelling part of this implementation is that anyone using this response builder from the reference implementation has immediate validation that their response is spec compliant or not.

Reply:
Thanks @amogh-jahagirdar for sharing your thoughts I feel like that was my understanding as well, the request and response require certain arguments to be present that the spec mandates, otherwise it is considered invalid and the `IllegalArgumentException` seems appropriate.

When checking in the iceberg's core `rest` module, we already do this in serval places so it seems to be the pattern.

Examples in the request classes
https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/rest/requests/CommitTransactionRequest.java#L40
https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/rest/requests/CreateNamespaceRequest.java#L48

Examples in parser classes
https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/rest/requests/CommitTransactionRequestParser.java#L44

https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java#L43

@rdblue If this is an issue though can try to follow your suggestion as well, let me know what you think.